wire-library/wire-protoc/src/main/java/com/squareup/wire/protocwire/Plugin.java

```diff
@@ -266,30 +184,67 @@ public static void run(CodeGenerator generator) throws PluginException {
    */
   public static void run(CodeGenerator generator, Environment environment) {
     CodeGeneratorRequest request;
+    ByteString rawRequest;
     try {
-      request = CodeGeneratorRequest.parseFrom(environment.getInputStream());
+      rawRequest = ByteString.readFrom(environment.getInputStream());
+      request = CodeGeneratorRequest.parseFrom(rawRequest);
     } catch (IOException e) {
-      throw new PluginException("protoc sent unparseable request to plugin.",
-        e);
+      throw new PluginException("protoc sent unparseable request to plugin.", e);
     }
 
-    List<FileDescriptor> filesToGenerate = parseFiles(request);
+    Map<String, FileDescriptor> files = asDescriptors(request.getProtoFileList());
+    ExtensionRegistry reg = createExtensionRegistry(files.values());
+
+    // now we must *re-parse* the request, but this time we can properly parse any
+    // custom options therein
+    try {
+      request = CodeGeneratorRequest.parseFrom(rawRequest, reg);
+    } catch (IOException e) {
+      throw new PluginException("protoc sent unparseable request to plugin.", e);
+    }
+    files = asDescriptors(request.getProtoFileList());
```

Comment:
I had to re-work all of this. The double-unmarshal is sadly needed to correctly interpret all custom options. Even though the code generator is working descriptor protos in the raw request, we do still need to "upgrade" them to proper linked descriptors in order to create the extension registry (for the second unmarshal) as well as to potentially be able to use custom options messages.